### ISSUE TITLE
Database Query unusual error handling

### ISSUE THREAD

**Issue by ProFire** [cakephp/cakephp#7743]
Hello bakers,

I recently encountered a problem with a database query error on CakePHP 3.1.4, an error that didn't occur in CakePHP 3.1.1. I believe this error is both my end and CakePHP's end. Took me hours to figure out where I went wrong.

### Error message encountered:
InvalidArgumentException: Cannot convert value to string in C:\wamp\www\SocialSoundfyr2\vendor\cakephp\cakephp\src\Database\Type\StringType.php

### Error trigger:
```php
$table = TableRegistry::get("ModelName");
return $table ->find("all", [
    "conditions" => [
        "ModelName.product_id" => $this->_properties["id"],
       "AssocModelName.status" => [0, 1],// This code worked in CakePHP 3.1.1
    ],
]);
```

### Solution:
```php
$table = TableRegistry::get("ModelName");
return $table ->find("all", [
    "conditions" => [
        "ModelName.product_id" => $this->_properties["id"],
       "AssocModelName.status IN" => [0, 1],// This code fixed the error
    ],
]);
```

### Debugging Procedures:
What I couldn't figure out is why the error message is totally unrelated to my mistake. Took me hours of tracing the error to find out where I did wrong. Here's some of my findings through my tracing:

**Level 1**
An array was passed to this class that's suppose to take in String type.
>Cake\Database\Type\StringType::toDatabase()
$value = [0,1];

**Level 2**
The type was determined to be string when the value is actually an Array type.
>Cake\Database\TypeConverterTrait::cast()
$value = [0,1]
$type = "string"

**Level 3**
The type passed to this class::function was null, thus at line 67, it was determined to be "string".
>Cake\Database\Statement\PDOStatement::bindValue()
$type = null

**Level 4**
The types array had a blank/null value for the specific key in the array.
>Cake\Database\Statement\StatementDecorator::bind()
$params = [2, [0,1]]
$types = ["integer", null]

### The Machine
OS: Windows 7
Stack: WAMP 64bit
PHP: 5.5.12
Database: Sqlite

**Comment by lorenzo** [comment]
I can try to improve the error message.

**Comment by markstory** [comment]
Closing as I think this was resolved in #7564
